packages/server/modules/core/services/objects/management.ts

```diff
@@ -0,0 +1,105 @@
+import crypto from 'crypto'
+import { RawSpeckleObject, SpeckleObject } from '@/modules/core/domain/objects/types'
+import { getMaximumObjectSizeMB } from '@/modules/shared/helpers/envHelper'
+import { estimateStringMegabyteSize } from '@/modules/core/utils/chunking'
+import { ObjectHandlingError } from '@/modules/core/errors/object'
+import { servicesLogger } from '@/logging/logging'
+import {
+  CreateObject,
+  StoreClosuresIfNotFound,
+  StoreSingleObjectIfNotFound
+} from '@/modules/core/domain/objects/operations'
+
+/**
+ * Note: we're generating the hash here, rather than on the db side, as there are
+ * limitations when doing upserts - ignored fields are not always returned, hence
+ * we cannot provide a full response back including all object hashes.
+ */
+export const prepInsertionObject = (
+  streamId: string,
+  obj: RawSpeckleObject
+): Omit<
+  SpeckleObject,
+  'totalChildrenCount' | 'totalChildrenCountByDepth' | 'createdAt'
+> => {
+  const MAX_OBJECT_SIZE_MB = getMaximumObjectSizeMB()
+
+  if (obj.hash) obj.id = obj.hash
+  else
+    obj.id =
+      obj.id || crypto.createHash('md5').update(JSON.stringify(obj)).digest('hex') // generate a hash if none is present
+
+  const stringifiedObj = JSON.stringify(obj)
+  const objectByteSize = estimateStringMegabyteSize(stringifiedObj)
+  if (objectByteSize > MAX_OBJECT_SIZE_MB) {
+    throw new ObjectHandlingError(
+      `Object too large. Object ID: ${obj.id}. (${objectByteSize} MB is > than limit, ${MAX_OBJECT_SIZE_MB} MB)`
+    )
+  }
+
+  return {
+    // Not sure why we stringify it, cause knex will do that, but leaving it as is for now
+    data: stringifiedObj as unknown as Record<string, unknown>,
```

Comment:
@didimitrie @gjedlicska Any clues on why we do this (see comment)? Usually for Jsonb columns the TS/JS data type is an object, cause Knex knows how to handle the objects and convert them to serialized JSON when inserting, and parsing the JSON when retrieving from the DB.

Is this some kind of preemptive optimization or?

Reply:
Same goes for totalChildrenCountByDepth below

Reply:
i'm seeing this used for obj size calculations later on, so i've just adjusted the insertion type to support strings